### PR TITLE
Improve Staking Limits

### PR DESCRIPTION
**Pull request by shawntabrizi**
When introducing the original staking limits: https://github.com/paritytech/substrate/pull/8920

we were a bit too strict on the behaviors allowed during otherwise okay scenarios.

The following changes are made in this PR:

* If we are not within our max nominator / validator threshold, we do not allow anyone to call `chill_other` on a different user.
* If we are an existing nominator / validator, we can still update our preferences using `nominate` / `validate`

polkadot companion: https://github.com/paritytech/polkadot/pull/3376

### CODE REVIEW

File: frame/staking/src/benchmarking.rs [hunk 1]
```diff
@@ -620,7 +620,11 @@ benchmarks! {
 		let (_, controller) = create_stash_controller::<T>(USER_SEED, 100, Default::default())?;
 		Staking::<T>::validate(RawOrigin::Signed(controller.clone()).into(), ValidatorPrefs::default())?;
 		Staking::<T>::update_staking_limits(
-			RawOrigin::Root.into(), BalanceOf::<T>::max_value(), BalanceOf::<T>::max_value(), None, None,
+			RawOrigin::Root.into(),
+			BalanceOf::<T>::max_value(),
+			BalanceOf::<T>::max_value(),
+			Some(0),
+			Some(0),
```
Comment (coriolinus) [inline frame/staking/src/benchmarking.rs:627]: This is because `Some(0)` forces a write, and `None` doesn't?

File: frame/staking/src/lib.rs [hunk 2]
```diff
@@ -2307,19 +2314,29 @@ pub mod pallet {
...
+			// and thus we are justified to chill this user.
 			//
 			// Otherwise, if caller is the same as the controller, this is just like `chill`.
 			if caller != controller {
 				let min_active_bond = if Nominators::<T>::contains_key(&stash) {
+					let max_nominator_count = MaxNominatorsCount::<T>::get().ok_or(Error::<T>::CannotChillOther)?;
+					let current_nominator_count = CounterForNominators::<T>::get();
+					// Safe for up to hundreds of millions of users.
+					let ratio = current_nominator_count.saturating_mul(100u32) / max_nominator_count.max(1u32);
+					ensure!(ratio > 75, Error::<T>::CannotChillOther);
```
Comment (coriolinus) [inline frame/staking/src/lib.rs:2328]: We should probably extract the 75 to a governance-controlled parameter. This seems like the sort of thing we'll want to be able to tune without requiring a chain upgrade.

File: frame/staking/src/lib.rs [hunk 3]
```diff
@@ -2307,19 +2314,29 @@ pub mod pallet {
...
 					MinNominatorBond::<T>::get()
 				} else if Validators::<T>::contains_key(&stash) {
+					let max_validator_count = MaxValidatorsCount::<T>::get().ok_or(Error::<T>::CannotChillOther)?;
+					let current_validator_count = CounterForValidators::<T>::get();
+					// Safe for up to hundreds of millions of users.
+					let ratio = current_validator_count.saturating_mul(100u32) / max_validator_count.max(1u32);
+					ensure!(ratio > 75, Error::<T>::CannotChillOther);
```
Comment (coriolinus) [inline frame/staking/src/lib.rs:2335]: Here also.

File: frame/staking/src/lib.rs [hunk 4]
```diff
@@ -2288,9 +2294,10 @@ pub mod pallet {
 		///
 		/// The dispatch origin for this call must be _Signed_, but can be called by anyone.
 		///
-		/// If the caller is the same as the controller being targeted, then no further checks
-		/// are enforced. However, this call can also be made by an third party user who witnesses
-		/// that this controller does not satisfy the minimum bond requirements to be in their role.
+		/// If the caller is the same as the controller being targeted, then no further checks are
+		/// enforced. However, this call can also be made by an third party user who witnesses that
+		/// the limits for number of validators or nominators is near the threshold (75%), and that
```
Comment (coriolinus) [inline frame/staking/src/lib.rs:2299]: Once the 75 is a named value, this should refer to the value instead of the literal.

**Review by kianenigma — APPROVED** [review]
+1 on the 75 being a param, otherwise LGTM

File: frame/staking/src/benchmarking.rs (hunk 1 above)
Comment (shawntabrizi) [reply frame/staking/src/benchmarking.rs:627]: No. 0 allows us to call chill other when none does not

File: frame/staking/src/lib.rs [hunk 5]
```diff
@@ -2307,19 +2314,29 @@ pub mod pallet {
 			let ledger = Self::ledger(&controller).ok_or(Error::<T>::NotController)?;
```
Comment (emostov) [inline frame/staking/src/lib.rs:2314]: Should the origin param for `chill_other` be configurable for the pallet? (I believe @shawntabrizi mentioned this idea already somewhere) The current, permission nature of this call is useful for scenarios where we want a network to be continuously "self adjusting". But I think we can imagine scenarios where this makes more sense as a governance only call (that then can be turned into a permission-less call with a runtime upgrade).

Comment (emostov) [reply frame/staking/src/lib.rs:2314]: Tagential, is it possible to make a call origin configurable without a runtime upgrade?

File: frame/staking/src/lib.rs [hunk 6]
```diff
@@ -2307,19 +2314,29 @@ pub mod pallet {
...
+			// If the caller is not the controller, we first check that we are near our threshold
+			// limit, and then we also check that the minimum bond requirements are not satisfied,
+			// and thus we are justified to chill this user.
 			//
 			// Otherwise, if caller is the same as the controller, this is just like `chill`.
 			if caller != controller {
 				let min_active_bond = if Nominators::<T>::contains_key(&stash) {
+					let max_nominator_count = MaxNominatorsCount::<T>::get().ok_or(Error::<T>::CannotChillOther)?;
+					let current_nominator_count = CounterForNominators::<T>::get();
+					// Safe for up to hundreds of millions of users.
```
Comment (emostov) [inline frame/staking/src/lib.rs:2326]: If we are concerned about saturating the multiplication we run into issues at tens of millions:

`2^32 -1 == 4_294_967_295`
`4_294_967_295 / 100 == 42_949_672`

So we would would saturate at `42_949_672` `current_nominator _count` - which seems totally fine, but just wanted to make sure I am understand what the comment is getting at

**Review by emostov — APPROVED** [review]
> +1 on the 75 being a param, otherwise LGTM

+1 :)

Outside the goals of this PR, I am curious if it is worth looking into a governance origin for `chill_other`; see https://github.com/paritytech/substrate/pull/9193#discussion_r658964909

File: frame/staking/src/lib.rs (hunk 5 above)
Comment (kianenigma) [reply frame/staking/src/lib.rs:2314]: > Tagential, is it possible to make a call origin configurable without a runtime upgrade?

Nope, that's logic change and needs altering `:code`.

**Comment by emostov** [comment]
Just read through commits f7c1698, 4723d5a, 1778970 and the updates lgtm 👍

**Review by coriolinus — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**